Feat/saves

lenafwu opened the pull request:
Added Saving Load manager, added ID to player, deck, ManaFuse bar

Added key support (F5 for saving, F9 for loading)

lenafwu committed:
add saving manager

lenafwu committed:
Merge remote-tracking branch 'origin/main' into feat/saves

lenafwu committed:
Update .gitignore

lenafwu committed:
Add save system integration with ISaveable interface for multiple game components, including PlayerManager, GameProgressTracker, and DeckManager. Implemented SaveLoadHandler for manual save/load operations and added PersistentId component to relevant prefabs. Updated Chamber_01 scene with new GameObject for SaveLoadManager.

gtaEPIC commented on .gitignore:
```diff
@@ -71,3 +71,6 @@ crashlytics-build.properties
 /[Aa]ssets/[Ss]treamingAssets/aa/*
+
+.sln
+.slnx
```
1. `.sln` is already ignored (`.slnx` is not)
2. Can someone explain why we would ignore it? I know Unity will just regenerate it but isn't this a file that probably shouldn't be ignored?

gtaEPIC replied:
```suggestion

*.slnx
```

Probably my only suggestion since `*.sln` already exists.

gtaEPIC reviewed (commented):
For the sake of cleanliness this is my only request to change, still going to test it as well but I think it's good.

gtaEPIC reviewed (changes requested):
Error logs:
```
Failed to load slot playersave: System.NullReferenceException: Object reference not set to an instance of an object
  at CardSystem.CardManager.SetCardInSlot (System.Int32 slotIndex, CardSystem.Card card) [0x00019] in C:\Users\REDACTED\Downloads\ManaFuse\Assets\Scripts\CardSystem\CardManager.cs:49 
  at CardSystem.DeckManager.RestoreState (System.Object state) [0x001ea] in C:\Users\REDACTED\Downloads\ManaFuse\Assets\Scripts\CardSystem\DeckManager.cs:364 
  at SaveSystem.SaveGameManager.Load (System.String slotName) [0x000f4] in C:\Users\REDACTED\Downloads\ManaFuse\Assets\Scripts\SaveSystem\SaveGameManager.cs:191 
UnityEngine.Debug:LogError (object)
SaveSystem.SaveGameManager:Load (string) (at Assets/Scripts/SaveSystem/SaveGameManager.cs:198)
SaveSystem.SaveLoadHandler:LoadGame () (at Assets/Scripts/SaveSystem/SaveLoadHandler.cs:81)
SaveSystem.SaveLoadHandler:Update () (at Assets/Scripts/SaveSystem/SaveLoadHandler.cs:36)
```
Savedata:
```
{
    "slotName": "playersave",
    "savedAtUtc": "2025-11-21T00:43:48.0503673Z",
    "states": [
        {
            "id": "8dd6dcf3e6b6418cbeccfbf62c3603d4",
            "type": "GameProgressTracker+ProgressSaveData, Assembly-CSharp, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null",
            "json": "{\"currentSceneIndex\":3,\"currentSceneName\":\"Chamber_01\",\"isVictoryAchieved\":false,\"enemiesRemaining\":7}"
        },
        {
            "id": "d33f9de9f22d4c6ab26b3428b8dc9058",
            "type": "CardSystem.DeckManager+DeckSaveData, Assembly-CSharp, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null",
            "json": "{\"baseDeckCardNames\":[\"Attack\",\"Attack\",\"Attack\",\"Attack_x3\",\"AttackBomb\"],\"activeDeckCardNames\":[\"Attack\"],\"discardPileCardNames\":[\"Attack\",\"Attack_x3\"],\"slotCardNames\":[\"\",\"\",\"AttackBomb\",\"Attack\"]}"
        },
        {
            "id": "a157675a29f54ceaa571e545283e0b21",
            "type": "PlayerManager+PlayerSaveData, Assembly-CSharp, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null",
            "json": "{\"manaUnits\":15,\"manaBarTicks\":1000,\"greenSliderValue\":544.3333129882813,\"redSliderValue\":544.3333129882813,\"positionX\":-13.495973587036133,\"positionY\":-1.600003719329834,\"positionZ\":4.128092288970947,\"roundedAngle\":0.0}"
        }
    ]
}
```

gtaEPIC commented on Assets/Scripts/CardSystem/DeckManager.cs:
```diff
@@ -279,5 +296,86 @@ public void AddTemporaryCardToDeck(List<Card> newCards)
+                    }
+                }
+
+                Debug.Log($"Deck state restored: {activeDeck.Count} active, {discardPile.Count} discarded");
```
Figured out the issue.
```suggestion
                // Restore cards in slots
                for (int i = 0; i < data.slotCardNames.Length && i < cardManager.cardSlots.Length; i++)
                {
                    if (!string.IsNullOrEmpty(data.slotCardNames[i]))
                    {
                        var card = baseDeck.FirstOrDefault(c => c.name == data.slotCardNames[i]);
                        if (card)
                        {
                            cardManager.SetCardInSlot(i, card);
                        }
                    }
                }

                Debug.Log($"Deck state restored: {activeDeck.Count} active, {discardPile.Count} discarded");
```

Sending a null card to the `SetCardInSlot` function will cause an error. My assumption is that this would be because there is supposed to be nothing in this slot, so telling it to be null seems wrong.

lenafwu committed:
Update .gitignore

Co-authored-by: Johnathan Hall <32531985+gtaEPIC@users.noreply.github.com>

lenafwu committed:
Update Assets/Scripts/CardSystem/DeckManager.cs

Co-authored-by: Johnathan Hall <32531985+gtaEPIC@users.noreply.github.com>

lenafwu committed:
Add null check in CardManager

lenafwu committed:
Fix the save/load for Deck

lenafwu commented:
@gtaEPIC Thanks for the review! I added .slnx since it keeps changing on my local setup...
I've committed your changes for the Deck/Card system, thanks for the update. Also updated the DeckManager and it should work now. 
Could you check if it works correctly now and see if it can be merged?

jpablo9988 committed:
Added Testing Scene and Script for Saving System

jpablo9988 reviewed (approved):
The save system seems to be working well.
There only change that needs to be done is connecting the SaveManager to GameContext instead of it being a singleton.
This, however, needs to be done when merging Alex's code refactor into the main. Right now, it works well as a modular component.

jpablo9988 committed:
Merge branch 'main' into feat/saves

jpablo9988 reviewed (changes requested):
The only change that I might do before merging is erasing any ISaveables from the live game itself as we haven't defined what we're actually saving and when. 
I pushed a Test Object we can test stuff with, and later we can implement when and what we want to save. 
We have to define this well as, because this game will eventually turn into a Rogue-like, stuff like Maximum Mana and Bullets in deck will be erased between playthroughs. The state of the actual scene, as they're short scenes, shouldn't be auto saved. We should only save at the beginning or end of a stage.

jpablo9988 commented:
> The only change that I might do before merging is erasing any ISaveables from the live game itself as we haven't defined what we're actually saving and when. I pushed a Test Object we can test stuff with, and later we can implement when and what we want to save. We have to define this well as, because this game will eventually turn into a Rogue-like, stuff like Maximum Mana and Bullets in deck will be erased between playthroughs. The state of the actual scene, as they're short scenes, shouldn't be auto saved. We should only save at the beginning or end of a stage.

I'll probably do this later, so don't worry too much about this @lenafwu.

lenafwu commented:
> The only change that I might do before merging is erasing any ISaveables from the live game itself as we haven't defined what we're actually saving and when. I pushed a Test Object we can test stuff with, and later we can implement when and what we want to save. We have to define this well as, because this game will eventually turn into a Rogue-like, stuff like Maximum Mana and Bullets in deck will be erased between playthroughs. The state of the actual scene, as they're short scenes, shouldn't be auto saved. We should only save at the beginning or end of a stage.

Thanks so much JP for helping with review and the explanations 👍! It makes sense now - I'll erase the saving for mana/deck. I'll see if I can add saving for the stages and I'll let you know if any question.

lenafwu committed:
Remove ISaveable in Deck/PlayerManager/GameProgressTracker
